openbsd/OpenBSDProcessList.c

```diff
@@ -290,14 +306,99 @@ void ProcessList_goThroughEntries(ProcessList* this) {
       }
 
       if (Process_isKernelThread(proc)) {
-         this->kernelThreads++;
+         this->super.kernelThreads++;
       }
 
-      this->totalTasks++;
+      this->super.totalTasks++;
       // SRUN ('R') means runnable, not running
       if (proc->state == 'P') {
-         this->runningTasks++;
+         this->super.runningTasks++;
       }
       proc->updated = true;
    }
 }
+
+static unsigned long long saturatingSub(unsigned long long a, unsigned long long b) {
+   return a > b ? a - b : 0;
+}
+
+static void getKernelCPUTimes(int cpuId, u_int64_t* times) {
+   int mib[] = { CTL_KERN, KERN_CPTIME2, cpuId };
+   size_t length = sizeof(u_int64_t) * CPUSTATES;
+   if (sysctl(mib, 3, times, &length, NULL, 0) == -1 ||
+         length != sizeof(u_int64_t) * CPUSTATES) {
+      CRT_fatalError("sysctl kern.cp_time2 failed");
+   }
+}
+
+static void kernelCPUTimesToHtop(const u_int64_t* times, CPUData* cpu) {
+   unsigned long long totalTime = 0;
+   for (int i = 0; i < CPUSTATES; i++) {
+      totalTime += times[i];
+   }
+
+   unsigned long long sysAllTime = times[CP_INTR] + times[CP_SYS];
+
+   // XXX Not sure if CP_SPIN should be added to sysAllTime.
+   // See https://github.com/openbsd/src/commit/531d8034253fb82282f0f353c086e9ad827e031c
+   #ifdef CP_SPIN
+   sysAllTime += times[CP_SPIN];
+   #endif
```

Comment:
I used `#ifdef CP_SPIN` everywhere because `CP_SPIN` is pretty much new and not available on 6.3.